core/modules.rs

```diff
@@ -299,39 +300,93 @@ impl RecursiveModuleLoad {
+  pub fn register_and_recurse(
     &mut self,
-    specifier: ModuleSpecifier,
-    referrer: ModuleSpecifier,
-  ) {
-    if !self.is_pending.contains(&specifier) {
-      let fut = self.loader.load(
-        self.op_state.clone(),
-        &specifier,
-        Some(referrer),
-        self.is_dynamic_import(),
+    scope: &mut v8::HandleScope,
+    module_map_rc: &Rc<RefCell<ModuleMap>>,
+    module_source: &ModuleSource,
+  ) -> Result<(), AnyError> {
+    // Register the module in the module map unless it's already there. If the
+    // specified URL and the "true" URL are different, register the alias.
+    if module_source.module_url_specified != module_source.module_url_found {
+      module_map_rc.borrow_mut().alias(
+        &module_source.module_url_specified,
+        &module_source.module_url_found,
       );
-      self.pending.push(fut.boxed_local());
-      self.is_pending.insert(specifier);
     }
+    let maybe_module_id = module_map_rc
+      .borrow()
+      .get_id(&module_source.module_url_found);
+    let module_id = match maybe_module_id {
+      Some(id) => {
+        debug!(
+          "Already-registered module fetched again: {}",
+          module_source.module_url_found
+        );
+        id
+      }
+      None => module_map_rc.borrow_mut().new_module(
+        scope,
+        self.is_currently_loading_main_module(),
+        &module_source.module_url_found,
+        &module_source.code,
+      )?,
+    };
+
+    // Recurse the module's imports. There are two cases for each import:
+    // 1. If the module is not in the module map, start a new load for it in
+    //    `self.pending`. The result of that load should eventually be passed to
+    //    this function for recursion.
+    // 2. If the module is already in the module map, queue it up to be
+    //    recursed synchronously here.
+    // This robustly ensures that the whole graph is in the module map before
+    // `LoadState::Done` is set.
+    let specifier =
+      crate::resolve_url(&module_source.module_url_found).unwrap();
+    let mut already_registered = VecDeque::new();
+    already_registered.push_back((module_id, specifier));
+    while let Some((module_id, referrer)) = already_registered.pop_front() {
+      let imports = module_map_rc
+        .borrow()
+        .get_children(module_id)
+        .unwrap()
+        .clone();
+      for specifier in imports {
+        if !self.visited.contains(&specifier) {
+          if let Some(module_id) =
+            module_map_rc.borrow().get_id(specifier.as_str())
+          {
+            already_registered.push_back((module_id, specifier.clone()));
+          } else {
+            let fut = self.loader.load(
+              self.op_state.clone(),
+              &specifier,
+              Some(referrer.clone()),
+              self.is_dynamic_import(),
+            );
+            self.pending.push(fut.boxed_local());
+          }
+          self.visited.insert(specifier);
+        }
+      }
+    }
```

Comment:
@bartlomieju See reworked code. Modules which are already in the module map are no longer reloaded, they are recursed eagerly and synchronously instead.